src/ray/core_worker/transport/direct_actor_task_submitter.cc

```diff
@@ -324,70 +332,68 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(
+                      rpc::ActorTableData::ActorState_Name(new_state) +
+                      "from disconnecting actor."));
 }
 
 void CoreWorkerDirectActorTaskSubmitter::CheckTimeoutTasks() {
-  auto task_info_list = std::make_shared<std::vector<TaskInfo>>();
+  // For each task in `wait_for_death_info_tasks`, if it times out, fail it with
+  // timeout_error_info. But operating on the queue requires the mu_ lock; while calling
+  // FailPendingTask requires the opposite. So we copy the tasks out from the queue within
+  // the lock. This requires putting the data into shared_ptr.
+  std::vector<std::shared_ptr<ClientQueue::PendingTaskWaitingForDeathInfo>> timeout_tasks;
+  int64_t now = current_time_ms();
   {
     absl::MutexLock lock(&mu_);
-    for (auto &queue_pair : client_queues_) {
-      auto &queue = queue_pair.second;
-      auto deque_itr = queue.wait_for_death_info_tasks.begin();
-      while (deque_itr != queue.wait_for_death_info_tasks.end() &&
-             /*timeout timestamp*/ deque_itr->first < current_time_ms()) {
-        auto &task_spec_status_pair = deque_itr->second;
-        task_info_list->push_back(TaskInfo{
-            task_spec_status_pair.first,
-            task_spec_status_pair.second,
-            queue_pair.first,
-            queue.preempted,
-        });
-        deque_itr = queue.wait_for_death_info_tasks.erase(deque_itr);
+    for (auto &[actor_id, client_queue] : client_queues_) {
+      auto &deque = client_queue.wait_for_death_info_tasks;
+      auto it = deque.begin();
+      while (it != deque.end()) {
+        if ((*it)->deadline_ms < now) {
+          (*it)->actor_preempted = client_queue.preempted;
+          timeout_tasks.push_back(*it);
+          it = deque.erase(it);
+        } else {
+          ++it;  // Only increment if not erasing
+        }
       }
     }
   }
-
-  if (task_info_list->empty()) {
-    return;
-  }
-
-  // Do not hold mu_, because FailPendingTask may call python from cpp,
-  // and may cause deadlock with SubmitActorTask thread when aquire GIL.
-  for (auto &task_info : *task_info_list) {
-    FailTaskWithError(task_info);
+  // Note: mu_ released.
+  for (auto &task : timeout_tasks) {
+    rpc::RayErrorInfo error_info;
+    if (!task->actor_preempted) {
+      error_info = task->timeout_error_info;
+    } else {
+      // special error for preempted actor. The task "timed out" because the actor may
+      // not have sent a notification to the gcs; regardless we already know it's
+      // preempted and it's dead.
+      rpc::ActorDeathCause &actor_death_cause = *error_info.mutable_actor_died_error();
+      actor_death_cause.mutable_actor_died_error_context()->set_actor_id(
+          task->task_spec.ActorId().Binary());
+      actor_death_cause.mutable_actor_died_error_context()->set_preempted(
+          task->actor_preempted);
+
+      error_info.set_error_type(rpc::ErrorType::ACTOR_DIED);
+      error_info.set_error_message("Actor died by preemption.");
```

Comment:
```suggestion
      error_info.set_error_message("Actor died by preemption.");
```

This error message is a bit unclear. Did the actor time out? Did we sigkill it?

Reply:
"by preemption" means the node is a spot instance and is preempted by the cloud provider. This is useful in some Anyscale production code. Previously it was `"Actor died."` which is even more vague :)